IBM Cloud Code Engine with MicroProfile/Open Liberty (#2616)
OpenLiberty/blogs: posts/2022-10-04-microprofile-serverless-ibm-code-engine.adoc

---
layout: post
title: "MicroProfile applications running as a serverless function via CloudEvent in IBM Cloud Code Engine"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/abutch3r
author_github: https://github.com/abutch3r
seo-title: MicroProfile applications running as a serverless function via CloudEvent in IBM Cloud Code Engine - OpenLiberty.io
seo-description: Running MicroProfile applications in the IBM Cloud Code Engine serverless environments using CloudEvents
blog_description: "Running MicroProfile applications in the IBM Cloud Code Engine serverless environments using CloudEvents"
---

> = MicroProfile applications running as a serverless function via CloudEvent in IBM Cloud Code Engine

Reviewer: ```suggestion
= MicroProfile applications running as serverless functions
```

Emily Jiang <https://github.com/Emily-Jiang>
Alex Butcher <https://github.com/abutch3r
:imagesdir: /
:url-prefix:
:url-about: /

Serverless is certainly a hot topic in 2022. MicroProfile has been adopted widely by cloud native applications. An often question being asked is:

Can MicroProfile Applications run as a serverless?

> This blog explains the basic concept of Serverless, CloudEvent and IBM Cloud Code Engine and then demonstrate how to get MicroProfile running on Open Liberty running as a serverless function in IBM Cloud Code Engine via CloudEvent.

Reviewer: ```suggestion
The short answer is YES. This blog explains the basic concept of Serverless, CloudEvent and IBM Cloud Code Engine and then demonstrate how to get an MicroProfile application running on Open Liberty running as a serverless function in IBM Cloud Code Engine via CloudEvent.
```

Let's first explain what is Serverless?

== Serverless
Serverless is a short name for Serverless Computing, which is an execution model in which the cloud provider allocates resources on demand. Serverless enables developers to concentrate on their own applications without the need of managing servers. It does not mean there is no server running in the background. On a contrary, normally Serverless architecture contains servers.

Serverless normally executes code in the form of a function, so Serverless sometimes referred to “Functions as a ServICCE” (FaaS). There are a few FaaS from major cloud providers:

*	IBM Cloud Code Engine: IBM Cloud Function
*	AWS: AWS Lambda
*	Microsoft Azure: Azure Functions
*	Google Cloud: Google Cloud Functions

> Serverless is a model. https://knative.dev/docs/concepts/[Knative] provides a platform-agnostic solution for running serverless deployment. CloudEvent is also a common event model to be used in Serverless architecture. This blog will focus on CloudEvent instead.

Reviewer: ```suggestion
Serverless is a model. In order to work with Serverless, Knative and Cloud Event are often used. https://knative.dev/docs/concepts/[Knative] provides a platform-agnostic solution for running serverless deployment. CloudEvent is also a common event model to be used in Serverless architecture. This blog will focus on CloudEvent instead.
```

== CloudEvents

> https://cloudevents.io/[CloudEvents] is a specification of describing event data in a common and language agnostic way. CloudEvents is to simplify event declaration and delivery across the servICCEs and platforms. This specification is under Cloud Native Computing Foundation (CNCF). CloudEvents are supported by many Serverless frameworks including IBM Cloud Code Engine.

Reviewer: ```suggestion
https://cloudevents.io/[CloudEvents] is a specification of describing event data in a common and language agnostic way. CloudEvents is to simplify event declaration and delivery across the services and platforms. This specification is under Cloud Native Computing Foundation (CNCF). CloudEvents are supported by many Serverless frameworks including IBM Cloud Code Engine, which will be discussed in the next section.
```

== IBM Cloud Code Engine

> https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine](ICCE) is a fully managed, serverless platform designed to run container images, batch jobs, or source code. Code Engine allows you to build event driven workloads that react to CloudEvents. Code Engine is truly serverless. It automatically scales your workloads up and down, and even down to zero when there are no requests. You only pay for the resources you consume. The common cloud native applications normal use MicroProfile specifications for configuring, securing, and observing applications.

Reviewer: ```suggestion
https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine](ICCE) is a fully managed, serverless platform designed to run container images, batch jobs, or source code. Code Engine allows you to build event driven workloads that react to CloudEvents. Code Engine is truly serverless. It automatically scales your workloads up and down, and even down to zero when there are no requests. You only pay for the resources you consume. You can easily run your cloud native applications in Code Engine. The common cloud native applications normal use MicroProfile specifications for configuring, securing, and observing applications.
```

== MicroProfile Applications
MicroProfile Applications are popular cloud native applications. They are designed for containers and clouds. Since we are entering into Serverless era, it will be very important to get MicroProfile Applications running in a serverless environment. Luckily with CloudEvent, it is very straightforward to get these kinds of applications running as functions in an environment that supports CloudEvents such as IBM Cloud Code Engine. In the next section, we will demonstrate how to get MicroProfile application deployed to Open Liberty running as a function in IBM Cloud Code Engine via CloudEvent.

== MicroProfile applications running as a serverless in Code Engine
RESTful MicroProfile Applications can run seamlessly as a serverless in Code Engine via CloudEvent. This section will explain in a step-by-step manner.

=== Pre-requisites
Before you start you will need the following

* IBM Cloud Account
* Docker
* Docker Registry that is externally accessible which can host your images (IBM Cloud Container Registry is available as part of your cloud account)

=== Creating your MicroProfile Application

> If you don't have a suitable RESTful application to host as a set of functions you can use https://openliberty.io/guides/getting-started.html[OpenLiberty and Microprofile Getting Started] to get a simple application well suited to a Serverless environment such as IBM Cloud Code Engine

Reviewer: ```suggestion
If you don't have a suitable RESTful application to host as a set of functions you can use https://openliberty.io/guides/getting-started.html[OpenLiberty and Microprofile Getting Started] to get a simple application well suited to a Serverless environment such as IBM Cloud Code Engine.
```

> It is recommended that if using the Getting Started code and you have experience with OpenLiberty and Microprofile that you use the `finish` folder as your starting point and go to the next step. If you do not, follow the guide up to the Running the application in a Docker container step.

Reviewer: ```suggestion
It is recommended that if using the Getting Started code and you have experience with OpenLiberty and MicroProfile that you use the `finish` folder as your starting point and go to the next step. If you do not, follow the guide up to the Running the application in a Docker container step.
```

* https://cloud.ibm.com/docs/codeengine?topic=codeengine-app-source-code[From source code respository] e.g. Github
* From a externally accessible Docker Registry:
** https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app[Public Registry] e.g. DockerHub
** https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-crimage[IBM Cloud Container Registry] - available as part of your IBM Cloud account
** https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-private[Private Registry]

Given to follow this blog, you will have an IBM Cloud Account, IBM Cloud Container Registry (ICCR) provides the quickest route to getting your image up and running.

=== Creating your IBM Cloud Code Engine Application
With your image uploaded, you now need to create and configure you IBM Cloud Code Engine Application to make your function available.

If you used ICCR follow these https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-crimage[steps to create your application using ICCR]. If your used one of the other types of registry follow the matching Application creating instructions.

When creating your application there are some considerations to make:

> * Image reference, while the name of your image will stay the same, you will be updating it later so you should consider whether to use the image hash that is has within the registry or if you want use the tag you uploaded with, just be aware that an update to the image may not be reflected when a new image with is pushed with the old tag.

Reviewer: ```suggestion
* Image reference, while the name of your image will stay the same, you will be updating it later so you should consider whether to use the image hash that is within the registry or if you want use the tag you uploaded with, just be aware that an update to the image may not be reflected when a new image with is pushed with the old tag.
```

author_github: https://github.com/abutch3r
seo-title: MicroProfile applications running as a serverless function via CloudEvent in IBM Cloud Code Engine - OpenLiberty.io
seo-description: Running MicroProfile applications in the IBM Cloud Code Engine serverless environments using CloudEvents
blog_description: "Running MicroProfile applications in the IBM Cloud Code Engine serverless functions"
additional_authors:
- name: Emily Jiang
  github: https://github.com/Emily-Jiang
  image: https://avatars0.githubusercontent.com/Emily-Jiang
---
= MicroProfile applications running as a serverless function via CloudEvent in IBM Cloud Code Engine
Alex Butcher <https://github.com/abutch3r>
:imagesdir: /
:url-prefix:
:url-about: /

Serverless is certainly a hot topic in 2022. MicroProfile has been adopted widely by cloud native applications. An often question being asked is:

Can MicroProfile Applications run as a serverless?

The short answer is YES. This blog explains the basic concept of Serverless, CloudEvent and IBM Cloud Code Engine and then demonstrate how to get an MicroProfile application running on Open Liberty running as a serverless function in IBM Cloud Code Engine via CloudEvent.

> Let's first explain what is Serverless?

Reviewer: ```suggestion
Let's first explain what is Serverless!
```

== Serverless

> Serverless is a short name for Serverless Computing, which is an execution model in which the cloud provider allocates resources on demand. Serverless enables developers to concentrate on their own applications without the need of managing servers. It does not mean there is no server running in the background. On a contrary, normally Serverless architecture contains servers.

Reviewer: ```suggestion
Serverless is a short name for Serverless Computing, which is an execution model in which the cloud provider allocates resources on demand. Serverless enables developers to concentrate on their own applications without the need of managing servers. It does not mean there is no server running in the background. On a contrary, Serverless architecture contains servers.
```

> Serverless normally executes code in the form of a function, so Serverless sometimes referred to “Functions as a ServICCE” (FaaS). There are a few FaaS from major cloud providers:

Reviewer: ```suggestion
Serverless normally executes code in the form of a function, so Serverless sometimes referred to “Functions as a Service” (FaaS) or "Container as a Serivce". There are a few major cloud providers supporting serverless:
```

== CloudEvents
https://cloudevents.io/[CloudEvents] is a specification of describing event data in a common and language agnostic way. CloudEvents is to simplify event declaration and delivery across the services and platforms. This specification is under Cloud Native Computing Foundation (CNCF). CloudEvents are supported by many Serverless frameworks including IBM Cloud Code Engine, which will be discussed in the next section.

== IBM Cloud Code Engine
https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine](ICCE) is a fully managed, serverless platform designed to run container images, batch jobs, or source code. Code Engine allows you to build event driven workloads that react to CloudEvents. Code Engine is truly serverless. It automatically scales your workloads up and down, and even down to zero when there are no requests. You only pay for the resources you consume. You can easily run your cloud native applications in Code Engine. The common cloud native applications normal use MicroProfile specifications for configuring, securing, and observing applications.

== MicroProfile Applications

> MicroProfile Applications are popular cloud native applications. They are designed for containers and clouds. Since we are entering into Serverless era, it will be very important to get MicroProfile Applications running in a serverless environment. Luckily with CloudEvent, it is very straightforward to get these kinds of applications running as functions in an environment that supports CloudEvents such as IBM Cloud Code Engine. In the next section, we will demonstrate how to get MicroProfile application deployed to Open Liberty running as a function in IBM Cloud Code Engine via CloudEvent.

Reviewer: ```suggestion
MicroProfile Applications are popular cloud native applications. MicroProfile offers a set of standards APIs for cloud-native applications that free your applications from vendor lock-in. Open Liberty is the leading implementation for MicroProfile specifications and we will use Open Liberty in this blog. 

Since we are entering into Serverless era, it will be very important to get MicroProfile Applications running in a serverless environment. Luckily with CloudEvent, it is very straightforward to get these kinds of applications running as functions in an environment that supports CloudEvents such as IBM Cloud Code Engine. In the next section, we will demonstrate how to get MicroProfile application deployed to Open Liberty running as a function in IBM Cloud Code Engine via CloudEvent.
```

If you make a request to `https://{ICCE_Application_URL}/` you should see the `Welcome to Open Liberty` page

To call the application we use the same path as used for the application locally
```
curl https://${ICCE_Application_URL}/dev/system/properties
```

After a short time you should see a response with a JSON payload containing all the system properties.

Note: All ICCE connections are HTTPS. so while `http` was used locally, the image is configured to support HTTPS without any changes.

Congratulations you have now called your application on IBM Cloud Code Engine.

While you are making the requests if you have the IBM Cloud Code Engine UI open on your applications Overview tab you will be able to see the active instances. you will need to be on the page when you start making the requests it creating the instance and then scaling down to zero when it stops receiving work

image:/img/blog/ibm-code-engine-active-instances.png[IBM Cloud Code Engine Active Instances,width=70%,align="center"]

=== Updating the application to use CloudEvents

> A common use case of serverless applications is to process events coming from non-http sources such as Kafka topics or Object Storage. Historically to consume these events the application had to make use of the Cloud Providers SDK, locking the application into that provider. This is where CloudEvents comes in, as it provides a cross-provider standard around which data can be transmitted and received by an application improving portability and reducing large dependencies.

Reviewer: Please move the above paragraph at line 130. and then add
`In this next section, we will demonstrate how to update your application to use CloudEvents.`
